src/ray/raylet/raylet_client.cc

```diff
@@ -0,0 +1,352 @@
+#include "raylet_client.h"
+
+#include <inttypes.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+
+#include "ray/common/common_protocol.h"
+#include "ray/ray_config.h"
+#include "ray/raylet/format/node_manager_generated.h"
+#include "ray/raylet/task_spec.h"
+#include "ray/util/logging.h"
+
+using MessageType = ray::protocol::MessageType;
+
+// TODO(rkn): The io methods below should be removed.
+int connect_ipc_sock(const std::string &socket_pathname) {
+  struct sockaddr_un socket_address;
+  int socket_fd;
+
+  socket_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+  if (socket_fd < 0) {
+    RAY_LOG(ERROR) << "socket() failed for pathname " << socket_pathname;
+    return -1;
+  }
+
+  memset(&socket_address, 0, sizeof(socket_address));
+  socket_address.sun_family = AF_UNIX;
+  if (socket_pathname.length() + 1 > sizeof(socket_address.sun_path)) {
+    RAY_LOG(ERROR) << "Socket pathname is too long.";
+    return -1;
```

Comment:
Here is resource leak. `close(socket_fd);` should be called before returning such as Line 44.